File: extension/su_mcp/tools/system.rb
module SU_MCP
  module Tools
    module System
      def self.ping(_params)
        { success: true, pong: true, timestamp: Time.now.to_i }
      end

      def self.undo(params)
        steps = [(params["steps"] || 1).to_i, 1].max
        model = Sketchup.active_model
        raise "No active model" unless model

        undone = 0
        steps.times do
          begin
            model.undo_operation
            undone += 1
          rescue StandardError
            break
          end
        end
        { success: true, undone: undone, requested: steps }
      end
    end
  end
end

SU_MCP::Dispatcher.register("ping")      { |params| SU_MCP::Tools::System.ping(params) }
SU_MCP::Dispatcher.register("undo_last") { |params| SU_MCP::Tools::System.undo(params) }

File: server/sketchup_mcp/tools/system.py
from mcp.server.fastmcp import FastMCP

from .. import connection


def register(mcp: FastMCP) -> None:
    @mcp.tool()
    def ping() -> dict:
        """Health check — round-trip to the SketchUp extension."""
        return connection.send("ping")

    @mcp.tool()
    def undo_last(steps: int = 1) -> dict:
        """Undo the last `steps` operations in the active model.

        Returns `{undone, requested}` — `undone` may be smaller if the
        undo stack is shorter than requested.
        """
        return connection.send("undo_last", {"steps": steps})

    @mcp.tool()
    def batch(
        calls: list[dict],
        wrap_undo: bool = True,
        stop_on_error: bool = True,
        undo_name: str = "MCP batch",
    ) -> dict:
        """Run multiple tool calls under a single undo operation.

        - `calls`: list of `{"tool": "name", "args": {...}}`
        - `wrap_undo`: wrap the whole batch in start_operation /
          commit_operation so it counts as one undo (default True)
        - `stop_on_error`: abort the batch (rolling back if wrapped)
          on the first failure (default True)

        Returns per-call results in `results[]` with `{index, success,
        result}` or `{index, success, error}`.
        """
        return connection.send("batch", {
            "calls": calls,
            "wrap_undo": wrap_undo,
            "stop_on_error": stop_on_error,
            "undo_name": undo_name,
        })
